specs/Standards/Tokens/NonFungibleToken/Event.md

```diff
@@ -134,12 +143,25 @@ EVENT_JSON:{
 }
 ```
 
+Contract metadata update:
+
+```js
+EVENT_JSON:{
+  "standard": "nep171",
+  "version": "1.1.0",
+  "event": "contract_metadata_update",
+  "data": [
+    { "metadata": { "base_uri": "..." } }
```

Comment:
I think it probably makes more sense to actually not include the `contract_metadata` at all in the event, primarily due to the fact that `icon` would be included which could easily put the log over the 16KB limit. I'll update to reflect this.

Reply:
@lachlanglen This is a very good point that I also was thinking about. I still think it is worth including the list of fields that were updated. What do you think?

Reply:
@frol I hear your point that the listener can choose to act only if the updated fields are relevant to them, but since they need to request the full metadata object anyway via `nft_metadata`, I'm not sure how much benefit this provides. Additionally, it introduces a requirement for the contract to keep track of which fields were updated, rather than simply being able to accept an `NFTContractMetadata` object in an update method and set the whole thing. Wdyt?

Reply:
@lachlanglen Are you familiar with the near-sdk-rs implementation of NFT events? There are two reasons for my question:
1. We will need someone to contribute the implementation to near-sdk-rs (Developer Governance strives for community contributions all across the ecosystem)
2. I wonder how to integrate this event in such a way, so implementors don't need to keep in mind that they need to emit the event (e.g. mint and burn events are emitted in the internal near-sdk-rs implementation)

Reply:
> but since they need to request the full metadata object anyway via nft_metadata, I'm not sure how much benefit this provides.

Well, fair point. Let's not go that route.

Reply:
Sure, I'm happy to contribute the near-sdk-rs implementation.

I guess we should add the update method then? thoughts on the naming? e.g. `contract_metadata_update` or `set_contract_metadata` (possibly more accurate if the whole object is being received and set rather than individual elements)

```
fn set_contract_metadata(&mut self, metadata: NFTContractMetadata);
```

Reply:
@lachlanglen Naming is hard, so I would like someone to find the best UX as part of this contribution. Currently, examples do not give any examples of updating the metadata, but implementors can alter the metadata directly, which should be redesigned, I believe.

https://github.com/near/near-sdk-rs/blob/master/examples/non-fungible-token/nft/src/lib.rs#L33

Reply:
@frol `implementors can alter the metadata directly` - what do you mean by this?

Reply:
@lachlanglen As NFT contract author I can just implement a new method like:

```rust
fn lets_update_metadata_and_do_something_else(&mut self) {
    self.metadata.set(NftMetadata { ... });
}
```

Given that `metadata` is defined as:

```rust
pub struct Contract {
    tokens: NonFungibleToken,
    metadata: LazyOption<NFTContractMetadata>,
}
```

The interface won't guard me against doing this, and this way I won't emit the metadata update event unless I know that I should emit it. This makes the interface fragile, and most of the implementations will miss this feature. Ideally, we should suggest using such an interface that would make it impossible to avoid emitting the event. (please, [see this talk](https://www.youtube.com/watch?v=3Q2hQfYW-XM))